File: cookbook/api-client.py
"""
api-client.py — CopilotForge Cookbook Recipe

WHAT THIS DOES:
    A typed async HTTP client wrapper using httpx. Includes auth header
    injection, automatic retry with exponential backoff, configurable
    timeouts, and structured response parsing with error handling.

WHEN TO USE THIS:
    When your project calls external APIs and you want consistent error
    handling, automatic retries, and typed responses with modern async Python.

HOW TO RUN:
    1. pip install httpx
    2. python cookbook/api-client.py

PREREQUISITES:
    - Python 3.10+
    - httpx >= 0.25.0
"""

from __future__ import annotations

import asyncio
import math
import os
import random
from dataclasses import dataclass, field
from typing import Any, TypeVar

import httpx

# --- Types ---

T = TypeVar("T")


@dataclass
class ApiClientConfig:
    base_url: str
    timeout_seconds: float = 30.0
    max_retries: int = 2
    retry_base_delay: float = 0.5
    default_headers: dict[str, str] = field(default_factory=dict)
    auth_token: str | None = None


@dataclass
class ApiResponse:
    """Typed wrapper around an HTTP response."""

    data: Any
    status_code: int
    headers: httpx.Headers


# --- Errors ---


class ApiError(Exception):
    """Raised when the API returns a non-2xx status code."""

    def __init__(self, message: str, status_code: int, body: Any) -> None:
        super().__init__(message)
        self.status_code = status_code
        self.body = body
        self.is_retryable = status_code == 429 or status_code >= 500


class ApiTimeoutError(Exception):
    """Raised when a request times out."""

    def __init__(self, url: str, timeout: float) -> None:
        super().__init__(f"Request to '{url}' timed out after {timeout}s")


# --- API Client ---


class ApiClient:
    """
    Async HTTP client with retry, auth, and structured error handling.

    Usage::

        async with ApiClient(ApiClientConfig(base_url="https://api.example.com")) as client:
            response = await client.get("/users")
            print(response.data)
    """

    def __init__(self, config: ApiClientConfig) -> None:
        self._config = config
        self._base_url = config.base_url.rstrip("/")
        self._client: httpx.AsyncClient | None = None

    async def __aenter__(self) -> "ApiClient":
        headers = {"Accept": "application/json", **self._config.default_headers}
        if self._config.auth_token:
            headers["Authorization"] = f"Bearer {self._config.auth_token}"

        self._client = httpx.AsyncClient(
            base_url=self._base_url,
            headers=headers,
            timeout=httpx.Timeout(self._config.timeout_seconds),
        )
        return self

    async def __aexit__(self, *exc: object) -> None:
        if self._client:
            await self._client.aclose()
            self._client = None

    # --- Public methods ---

    async def get(self, path: str, *, params: dict[str, Any] | None = None) -> ApiResponse:
        return await self._request("GET", path, params=params)

    async def post(self, path: str, *, json: Any = None, params: dict[str, Any] | None = None) -> ApiResponse:
        return await self._request("POST", path, json=json, params=params)

    async def put(self, path: str, *, json: Any = None, params: dict[str, Any] | None = None) -> ApiResponse:
        return await self._request("PUT", path, json=json, params=params)

    async def patch(self, path: str, *, json: Any = None, params: dict[str, Any] | None = None) -> ApiResponse:
        return await self._request("PATCH", path, json=json, params=params)

    async def delete(self, path: str, *, params: dict[str, Any] | None = None) -> ApiResponse:
        return await self._request("DELETE", path, params=params)

    # --- Core request logic ---

    async def _request(
        self,
        method: str,
        path: str,
        *,
        json: Any = None,
        params: dict[str, Any] | None = None,
    ) -> ApiResponse:
        if self._client is None:
            raise RuntimeError("ApiClient must be used as an async context manager")

        last_error: BaseException | None = None

        for attempt in range(self._config.max_retries + 1):
            try:
                response = await self._client.request(
                    method, path, json=json, params=params
                )

                # Parse response body.
                content_type = response.headers.get("content-type", "")
                data: Any
                if "application/json" in content_type:
                    data = response.json()
                else:
                    data = response.text

                # Non-2xx status — raise ApiError.
                if not response.is_success:
                    raise ApiError(
                        f"{method} {path} returned {response.status_code}",
                        response.status_code,
                        data,
                    )

                return ApiResponse(
                    data=data,
                    status_code=response.status_code,
                    headers=response.headers,
                )

            except httpx.TimeoutException:
                raise ApiTimeoutError(
                    f"{self._base_url}{path}", self._config.timeout_seconds
                ) from None

            except ApiError as exc:
                last_error = exc
                is_last = attempt >= self._config.max_retries

                if not exc.is_retryable or is_last:
                    raise

                delay = self._config.retry_base_delay * (2**attempt) + random.random() * self._config.retry_base_delay
                print(
                    f"  [api-client] Retry {attempt + 1}/{self._config.max_retries} "
                    f"for {method} {path} in {delay:.1f}s"
                )
                await asyncio.sleep(delay)

        # Unreachable, but satisfies the type checker.
        raise last_error  # type: ignore[misc]


# --- Example Usage ---


async def main() -> None:
    print("=== API Client Recipe ===\n")

    # TODO: Replace with your actual API base URL and auth token.
    config = ApiClientConfig(
        base_url="https://jsonplaceholder.typicode.com",
        timeout_seconds=10.0,
        max_retries=2,
        auth_token=os.environ.get("API_TOKEN", "demo-token"),
    )

    async with ApiClient(config) as client:
        try:
            # GET with query params.
            print("1. GET /todos (first 3):")
            resp = await client.get("/todos", params={"_limit": 3})
            for todo in resp.data:
                check = "✅" if todo["completed"] else "⬜"
                print(f"   {check} [{todo['id']}] {todo['title']}")

            # POST request.
            print("\n2. POST /todos:")
            resp = await client.post(
                "/todos",
                json={"title": "Write cookbook recipe", "completed": False, "userId": 1},
            )
            print(f"   Created todo #{resp.data['id']} (status: {resp.status_code})")

            # GET single item.
            print("\n3. GET /todos/1:")
            resp = await client.get("/todos/1")
            status = "done" if resp.data["completed"] else "pending"
            print(f"   {resp.data['title']} — {status}")

        except ApiError as exc:
            print(f"API error [{exc.status_code}]: {exc}")
        except ApiTimeoutError as exc:
            print(f"Timeout: {exc}")
        except Exception as exc:
            print(f"Unexpected error: {exc}")
            raise


if __name__ == "__main__":
    asyncio.run(main())

File: cookbook/api-client.ts
/**
 * api-client.ts — CopilotForge Cookbook Recipe
 *
 * WHAT THIS DOES:
 *   A typed HTTP client wrapper using native fetch. Includes auth header
 *   injection, automatic retry with exponential backoff, configurable
 *   timeouts, and structured response parsing with error handling.
 *
 * WHEN TO USE THIS:
 *   When your project calls external APIs and you want consistent error
 *   handling, automatic retries, and typed responses without adding axios
 *   or other HTTP libraries.
 *
 * HOW TO RUN:
 *   1. npx ts-node cookbook/api-client.ts
 *
 * PREREQUISITES:
 *   - Node.js 18+ (native fetch required)
 *   - TypeScript 5+
 */

// --- Types ---

interface ApiClientConfig {
  baseUrl: string;
  /** Default timeout in milliseconds. */
  timeoutMs?: number;
  /** Maximum retry attempts for transient failures. */
  maxRetries?: number;
  /** Base delay for exponential backoff (ms). */
  retryBaseDelayMs?: number;
  /** Default headers applied to every request. */
  defaultHeaders?: Record<string, string>;
  /** Called before every request — return headers to merge. */
  getAuthHeaders?: () => Promise<Record<string, string>> | Record<string, string>;
}

interface ApiRequestOptions {
  method?: "GET" | "POST" | "PUT" | "PATCH" | "DELETE";
  headers?: Record<string, string>;
  body?: unknown;
  timeoutMs?: number;
  /** Override the default retry count for this request. */
  maxRetries?: number;
  /** Query parameters appended to the URL. */
  params?: Record<string, string | number | boolean>;
}

interface ApiResponse<T> {
  data: T;
  status: number;
  headers: Headers;
}

// --- Errors ---

class ApiError extends Error {
  readonly status: number;
  readonly body: unknown;
  readonly isRetryable: boolean;

  constructor(message: string, status: number, body: unknown) {
    super(message);
    this.name = "ApiError";
    this.status = status;
    this.body = body;
    // 429 (rate limit) and 5xx are retryable.
    this.isRetryable = status === 429 || status >= 500;
  }
}

class ApiTimeoutError extends Error {
  constructor(url: string, timeoutMs: number) {
    super(`Request to '${url}' timed out after ${timeoutMs}ms`);
    this.name = "ApiTimeoutError";
  }
}

// --- API Client ---

class ApiClient {
  private config: Required<
    Pick<ApiClientConfig, "baseUrl" | "timeoutMs" | "maxRetries" | "retryBaseDelayMs" | "defaultHeaders">
  > & { getAuthHeaders?: ApiClientConfig["getAuthHeaders"] };

  constructor(config: ApiClientConfig) {
    this.config = {
      baseUrl: config.baseUrl.replace(/\/+$/, ""),
      timeoutMs: config.timeoutMs ?? 30_000,
      maxRetries: config.maxRetries ?? 2,
      retryBaseDelayMs: config.retryBaseDelayMs ?? 500,
      defaultHeaders: config.defaultHeaders ?? {},
      getAuthHeaders: config.getAuthHeaders,
    };
  }

  // --- Public methods ---

  async get<T>(path: string, options?: Omit<ApiRequestOptions, "method" | "body">): Promise<ApiResponse<T>> {
    return this.request<T>(path, { ...options, method: "GET" });
  }

  async post<T>(path: string, body?: unknown, options?: Omit<ApiRequestOptions, "method" | "body">): Promise<ApiResponse<T>> {
    return this.request<T>(path, { ...options, method: "POST", body });
  }

  async put<T>(path: string, body?: unknown, options?: Omit<ApiRequestOptions, "method" | "body">): Promise<ApiResponse<T>> {
    return this.request<T>(path, { ...options, method: "PUT", body });
  }

  async patch<T>(path: string, body?: unknown, options?: Omit<ApiRequestOptions, "method" | "body">): Promise<ApiResponse<T>> {
    return this.request<T>(path, { ...options, method: "PATCH", body });
  }

  async delete<T>(path: string, options?: Omit<ApiRequestOptions, "method" | "body">): Promise<ApiResponse<T>> {
    return this.request<T>(path, { ...options, method: "DELETE" });
  }

  // --- Core request logic ---

  private async request<T>(path: string, options: ApiRequestOptions = {}): Promise<ApiResponse<T>> {
    const method = options.method ?? "GET";
    const url = this.buildUrl(path, options.params);
    const timeoutMs = options.timeoutMs ?? this.config.timeoutMs;
    const maxRetries = options.maxRetries ?? this.config.maxRetries;

    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      Accept: "application/json",
      ...this.config.defaultHeaders,
    };

    // Inject auth headers.
    if (this.config.getAuthHeaders) {
      const authHeaders = await this.config.getAuthHeaders();
      Object.assign(headers, authHeaders);
    }

    if (options.headers) {
      Object.assign(headers, options.headers);
    }

    let lastError: unknown;

    for (let attempt = 0; attempt <= maxRetries; attempt++) {
      try {
        const response = await this.fetchWithTimeout(url, {
          method,
          headers,
          body: options.body !== undefined ? JSON.stringify(options.body) : undefined,
        }, timeoutMs);

        // Parse response body.
        const contentType = response.headers.get("content-type") ?? "";
        let data: unknown;

        if (contentType.includes("application/json")) {
          data = await response.json();
        } else {
          data = await response.text();
        }

        // Non-2xx status — throw an ApiError.
        if (!response.ok) {
          throw new ApiError(
            `${method} ${path} returned ${response.status}`,
            response.status,
            data,
          );
        }

        return { data: data as T, status: response.status, headers: response.headers };
      } catch (error) {
        lastError = error;

        const isRetryable =
          error instanceof ApiError ? error.isRetryable :
          error instanceof ApiTimeoutError ? true :
          false;

        const isLastAttempt = attempt >= maxRetries;

        if (!isRetryable || isLastAttempt) {
          throw error;
        }

        // Exponential backoff with jitter.
        const delay = this.config.retryBaseDelayMs * Math.pow(2, attempt) + Math.random() * this.config.retryBaseDelayMs;
        console.log(`  [api-client] Retry ${attempt + 1}/${maxRetries} for ${method} ${path} in ${Math.round(delay)}ms`);
        await new Promise((resolve) => setTimeout(resolve, delay));
      }
    }

    throw lastError;
  }

  private async fetchWithTimeout(url: string, init: RequestInit, timeoutMs: number): Promise<Response> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeoutMs);

    try {
      return await fetch(url, { ...init, signal: controller.signal });
    } catch (error) {
      if (error instanceof DOMException && error.name === "AbortError") {
        throw new ApiTimeoutError(url, timeoutMs);
      }
      throw error;
    } finally {
      clearTimeout(timer);
    }
  }

  private buildUrl(path: string, params?: Record<string, string | number | boolean>): string {
    const url = new URL(`${this.config.baseUrl}${path}`);
    if (params) {
      for (const [key, value] of Object.entries(params)) {
        url.searchParams.set(key, String(value));
      }
    }
    return url.toString();
  }
}

// --- Example Usage ---

// Types for the example API.
interface Todo {
  id: number;
  title: string;
  completed: boolean;
  userId: number;
}

async function main() {
  console.log("=== API Client Recipe ===\n");

  // TODO: Replace with your actual API base URL and auth.
  const client = new ApiClient({
    baseUrl: "https://jsonplaceholder.typicode.com",
    timeoutMs: 10_000,
    maxRetries: 2,
    getAuthHeaders: async () => ({
      // TODO: Replace with real token retrieval.
      Authorization: `Bearer ${process.env.API_TOKEN ?? "demo-token"}`,
    }),
  });

  try {
    // GET request with query params.
    console.log("1. GET /todos (first 3):");
    const { data: todos } = await client.get<Todo[]>("/todos", {
      params: { _limit: 3 },
    });
    for (const todo of todos) {
      console.log(`   ${todo.completed ? "✅" : "⬜"} [${todo.id}] ${todo.title}`);
    }

    // POST request.
    console.log("\n2. POST /todos:");
    const { data: created, status } = await client.post<Todo>("/todos", {
      title: "Write cookbook recipe",
      completed: false,
      userId: 1,
    });
    console.log(`   Created todo #${created.id} (status: ${status})`);

    // GET single item.
    console.log("\n3. GET /todos/1:");
    const { data: single } = await client.get<Todo>("/todos/1");
    console.log(`   ${single.title} — ${single.completed ? "done" : "pending"}`);
  } catch (error) {
    if (error instanceof ApiError) {
      console.error(`API error [${error.status}]:`, error.message);
    } else if (error instanceof ApiTimeoutError) {
      console.error("Timeout:", error.message);
    } else {
      console.error("Unexpected error:", error);
      throw error;
    }
  }
}

main().catch(console.error);
